File: src/crypto_tools.js
const crypto = require('crypto')
/**------------------
 * @enum Mode
 * ------------------
 * @description AES encryption modes
 *  AES can be use multiple approach, so you can use all supported mode in this list
 * --------------------------
 */
const Mode = {
    'CBC': 'aes-256-cbc',
    'CFB': 'aes-256-cfb',
    'CTR': 'aes-256-ctr',
    'ECB': 'aes-256-ecb',
    'OFB': 'aes-256-ofb'
}
/**------------------
 * @class AESCipher
 * @param {string} key
 * @param {Mode} mode
 * ------------------
 * @description AES encryption class
 * ------------------------------
 */
class AESCipher {

    /**------------ 
     * @constructor
     * @param {string} key
     * @param {Mode} mode
     * ------------
    */
    constructor(key, mode) {
        const supportedModes = [Mode.CBC, Mode.CFB, Mode.CTR, Mode.ECB, Mode.OFB];
        if (!supportedModes.includes(mode)) {
            throw new Error("Mode not supported");
        }
        this.key = key
        this.mode = mode;
    }


    /**------------------
     * @function encrypt
     * @param {Buffer} data
     * @returns {Buffer} // encrypted data
     * ------------------
     * @description encrypt data
     * --------------------------
     */
    encrypt(data) {
        // iV is generated for each encryption for key security.
        // iv vector is added beginning of encrypted data. 
        // It's not used for data, used for key
        const initVector = crypto.randomBytes(16);

        // buffer converting base64 string
        const base64_data = data.toString("base64");
        //Note: block-based cipher algorithms use padding. 
        //      Padding fills this raw data to the block size, 
        // making it workable by the encryption algorithm.
        //      Padding size information is added end of the data. 
        //      When the decrypt the data, the padding section should be removed 
        // as much as the padding size by looking end of the data 
        // creating cipher. in this crypto library autoamtically handle padding.
        const cipher = crypto.createCipheriv(this.mode, this.key, initVector);

        // encrypt
        let encryptedData = cipher.update(base64_data, "base64", "base64");
        encryptedData += cipher.final("base64");
        // encrypted base64 string data to buffer
        const Combine_Buffer = Buffer.concat([initVector, Buffer.from(encryptedData, "base64")]);
        return Combine_Buffer;
    }

    /**------------------
     * @function decrypt
     * @param {Buffer} data
     * @returns {Buffer} // decrypted data
     * ------------------
     */
    decrypt(data) {

        // Iv vector is read in encrypted data
        const initVector = data.slice(0, 16);
        //Note: block-based cipher algorithms use padding. 
        //      Padding fills this raw data to the block size, 
        // making it workable by the encryption algorithm.
        //      Padding size information is added end of the data. 
        //      When the decrypt the data, the padding section should be removed 
        // as much as the padding size by looking end of the data 
        // creating decipher. in this crypto library autoamtically handle padding.
        const decipher = crypto.createDecipheriv(this.mode, this.key, initVector);
        // decrypt
        let decryptedData = decipher.update(data.slice(16), "base64", "base64");
        decryptedData += decipher.final("base64");
        
        // decrypted base64 string data to buffer
        const decrypted_Buffer = Buffer.from(decryptedData, "base64");
        return decrypted_Buffer;
    };

    /**------------------
     * @function checkEncryption
     * @param {Buffer} encrypted
     * @param {Buffer} data
     * @returns {boolean}
     * ------------------
     * @description check if the encrypted data is the same as the original data
     * it takes decrypted buffer and original buffer and compare.
     * --------------------------
     */
    checkEncryption(decrypted, data) {
        if (data.equals(decrypted)) {
            return true;
        }
        else {
            return false;
        }
    }
}

module.exports = {
    AESCipher,
    Mode
}

File: src/crypto_tools.py
from Crypto.Cipher import AES
from Crypto.Util.Padding import pad, unpad
from enum import Enum
import os
class Modes(Enum):
    ECB = 1        #: Electronic Code Book (:ref:`ecb_mode`)
    CBC = 2        #: Cipher-Block Chaining (:ref:`cbc_mode`)
    CFB = 3        #: Cipher Feedback (:ref:`cfb_mode`)    
    OFB = 5        #: Output Feedback (:ref:`ofb_mode`)
    CTR = 6        #: Counter mode (:ref:`ctr_mode`)
    OPENPGP = 7    #: OpenPGP mode (:ref:`openpgp_mode`)
    CCM = 8        #: Counter with CBC-MAC (:ref:`ccm_mode`)
    EAX = 9        #: :ref:`eax_mode`
    SIV = 10       #: Galois Counter Mode (:ref:`gcm_mode`)
    GCM = 11       #: Synthetic Initialization Vector (:ref:`siv_mode`)
    OCB = 12       #: Offset Code Book (:ref:`ocb_mode`)


class AESCipher:

    """
    @constructor
    @param key : 256 bit string key 
    @param mode: Aes algorithm modes   
    """ 
    def __init__(self, key, mode):
        self.key = key
        self.mode = mode.value
        self.block_size = 16
        
    """
    @param data: bytes
    @description: encrypt data
    @return: bytes
    """
    def encrypt(self, data):
        iv = os.urandom(self.block_size)
        cipher = AES.new(self.key, self.mode,iv)
        padded_data = pad(data, self.block_size)
        cipher_text = cipher.encrypt(padded_data)

        # merge bytes 
        iv_and_cipher_text = iv + cipher_text
                
        return iv_and_cipher_text


    """
    @param data: bytes
    @description: decrypt data
    @return: bytes
    """
    def decrypt(self, data):
        
        iv = data[:self.block_size];
        
         # get cipher text
        encypted_text = data[self.block_size:]
        
        # create cipher object
        cipher = AES.new(self.key, self.mode, iv)
        
        # decrypt data
        padded_plain_text = cipher.decrypt(encypted_text)
        
        # unpad data
        plain_text = unpad(padded_plain_text, self.block_size)
        # return plain text
        return plain_text
    
    """
    @param origin_data: bytes
    @param decrypted_data: bytes
    return succes: boolean 
    """
    def checkEncryption(self, origin_data, decrypted_data):
        if len(origin_data) != len(decrypted_data):
            return False
        for i in range(len(decrypted_data)):
            if decrypted_data[i] != origin_data[i]:
                return False
        return True
